storage/innobase/row/row0merge.cc

```diff
@@ -5370,12 +5370,21 @@ void trx_t::bulk_rollback_low()
 
 dberr_t trx_t::bulk_insert_apply_for_table(dict_table_t *table)
 {
-  auto t= check_bulk_buffer(table);
-  if (!t || !t->is_bulk_insert())
+  if (UNIV_UNLIKELY(!bulk_insert))
     return DB_SUCCESS;
-  dberr_t err= t->write_bulk(table, this);
-  if (err != DB_SUCCESS)
-    bulk_rollback_low();
+  ut_ad(!check_unique_secondary);
+  ut_ad(!check_foreigns);
+  auto it= mod_tables.find(table);
+  if (it == mod_tables.end()) return DB_SUCCESS;
+
+  dberr_t err= DB_SUCCESS;
+  if (it->second.bulk_store)
+  {
+    err= it->second.write_bulk(table, this);
+    if (err) bulk_rollback_low();
+  }
+  if (it->second.is_bulk_insert())
+    it->second.end_bulk_insert();
   return err;
```

Comment:
If I understood it correctly, the last `if` is unnecessary and we can just unconditionally invoke `end_bulk_insert()`, which will execute `first&= ~BULK`. It is better to avoid unnecessary conditions.

Could `trx_t::bulk_rollback_low()` execute `t.second.end_bulk_insert()`? That would seem to be more robust to me. In that case, this could be simplified to the following:
```c++
  auto it= mod_tables.find(table);
  if (it != mod_tables.end() && it->second.bulk_store)
    if (dberr_t err= it->second.write_bulk(table, this))
      return err;
  it->second.end_bulk_insert();
  return DB_SUCCESS;
```